### PR TITLE
Optimize Wan VAE: Add Temporal Chunking, On-TPU Postprocessing, Static JIT Support, and Configurable Dtypes

### PR DESCRIPTION
**Pull request by Perseus14**
## Overview
This pull request introduces significant memory, compilation, and bandwidth optimizations for the Wan (T2V, I2V, Vace and Animate) VAE encoding and decoding execution graphs on TPUs. By enforcing static shapes and moving postprocessing operations on-device, this PR significantly improves scaling efficiency and execution robustness.

### 1. VAE Temporal Chunking & Static Compilation
* Introduced configurable chunk size parameters (`vae_decode_chunk` and `vae_encode_chunk`) in the base configuration files.
* Added static temporal zero-padding prior to `jax.lax.scan` loops and precise output trimming post-scan. This guarantees uniform static shapes across iterations, allowing seamless static JIT/XLA compilation without dynamic shape fallbacks or recompilation overhead.

### 2. On-TPU Output Quantization
* Transitioned the final video post-processing stages (range rescaling to `[0, 1]`, clipping, and scaling to `uint8`) directly into the JAX/TPU execution graph across all Wan pipelines (`WanPipeline`, `WanPipelineI2V`, and `WanAnimatePipeline`).
* Gathering sharded arrays as `uint8` instead of floating-point cuts device-to-host memory interconnect transfer bandwidth by up to 4x.
* Completely eliminated the need for `torch.Tensor` conversions and PyTorch processor overhead during generation output assembly.

### 3. Distributed Sharding Robustness
* Standardized VAE 1D spatial sharding rules to use `P("redundant", None, None, "vae_spatial", None)`. Explicitly marking the batch axis as redundant prevents unintended cross-mesh synchronization across replicas under XLA SPMD partitioning.
* Added defensive dimension existence checks when accessing mesh properties in `attention_flax.py` to support diverse hardware topologies gracefully.

### 4. Configurable Activation & Weight Precision
* Integrated robust type casting in `pyconfig.py` for user-defined numerical precision (`vae_dtype`, `vae_weights_dtype`, and `scheduler_dtype`), defaulting VAE runtime memory down to 16-bit precision to save 50% High Bandwidth Memory (HBM) capacity.

| Model | Variant | BaselIne Generation Time | Current Generation Time |
|---|---|---|---|
| WAN2.2 | T2V | 132.2s | 130.6s |
| WAN2.2 | I2V | 133.3s | 132.3s |
| WAN2.1 | T2V | 132.1s | 130.6s |
| WAN2.1 | I2V | 142.7s | 141.6s |

**Conclusion**: No visual change from baseline across all tested variants.

### CODE REVIEW

File: src/maxdiffusion/pipelines/wan/wan_pipeline_animate.py [hunk 1]
```diff
@@ -1041,5 +1041,13 @@ def __call__(
       return seg_latents
 
     # Postprocess to [0, 1] numpy.
-    video_torch = torch.from_numpy(np.array(video_cf.astype(jnp.float32))).to(torch.bfloat16)
-    return self.video_processor.postprocess_video(video_torch, output_type="np")
+    with self.vae_mesh, nn_partitioning.axis_rules(self.vae_logical_axis_rules):
```
Comment (ninatu) [inline src/maxdiffusion/pipelines/wan/wan_pipeline_animate.py:1049]: Would it be possible to use self._decode_latents_to_video here?

File: src/maxdiffusion/pipelines/wan/wan_pipeline_2_1.py [hunk 2]
```diff
@@ -176,7 +176,7 @@ def __call__(
     trace["denoise_total"] = time.perf_counter() - t_denoise_start
 
     t_decode_start = time.perf_counter()
-    video = self._decode_latents_to_video(latents)
+    video = self._decode_latents_to_video(latents, trace=trace)
```
Comment (ninatu) [inline src/maxdiffusion/pipelines/wan/wan_pipeline_2_1.py:181]: Could it be also added in Wan-VACE pipeline, that all pipelines are aligned? Now all pipelines except VACE return np.array in [0, 225] and only VACE returns in [0. 1]

File: src/maxdiffusion/configs/base_wan_14b.yml [hunk 3]
```diff
@@ -41,10 +41,19 @@ revision: ''
 weights_dtype: 'bfloat16'
 # This sets the layer's dtype in the model. Ex: nn.Dense(dtype=activations_dtype)
 activations_dtype: 'bfloat16'
+vae_weights_dtype: 'bfloat16'
```
Comment (ninatu) [inline src/maxdiffusion/configs/base_wan_14b.yml]: Could default vae_weights_dtype and vae_dtype be kept to float32 as in the original models? Since bfloat16  might add unexpected artifacts. It refers to all configs

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py [hunk 4]
```diff
@@ -1111,64 +1141,98 @@ def _encode(self, x: jax.Array, feat_cache: AutoencoderKLWanCache):
...
     t = x.shape[1]
-    iter_ = 1 + (t - 1) // 4
+    CHUNK_SIZE = self.vae_encode_chunk
+    if CHUNK_SIZE == -1:
+      CHUNK_SIZE = max(1, t - 1)
+      assert (
+          CHUNK_SIZE % self.temporal_downsample_factor == 0
+      ), f"When vae_encode_chunk is -1, the input sequence length - 1 ({CHUNK_SIZE}) must be a multiple of the temporal downsampling factor ({self.temporal_downsample_factor})."
+    # Number of chunk iterations: 1 for init frame, then ceil((t-1)/CHUNK_SIZE) for the rest
+    iter_ = 1 + ((t - 1 + CHUNK_SIZE - 1) // CHUNK_SIZE) if t > 1 else 1
```
Comment (ninatu) [inline src/maxdiffusion/models/wan/autoencoder_kl_wan.py:1151]: I'm not sure that this logic is correct, shouldn't it be iter_ = 1 + ((t - 1) // CHUNK_SIZE) ?

Comment (Perseus14) [reply src/maxdiffusion/models/wan/autoencoder_kl_wan.py:1151]: Thanks, done!

File: src/maxdiffusion/configs/base_wan_14b.yml (hunk 3 above)
Comment (Perseus14) [reply src/maxdiffusion/configs/base_wan_14b.yml]: Done!

File: src/maxdiffusion/pipelines/wan/wan_pipeline_2_1.py (hunk 2 above)
Comment (Perseus14) [reply src/maxdiffusion/pipelines/wan/wan_pipeline_2_1.py:181]: Done!

File: src/maxdiffusion/pipelines/wan/wan_pipeline_animate.py (hunk 1 above)
Comment (Perseus14) [reply src/maxdiffusion/pipelines/wan/wan_pipeline_animate.py:1049]: For `WANAnimate` we use `self._decode_segment_to_pixels` as inference runs over multi-frame sequential overlap segments and each segment needs to be decoded immediately into raw pixel space to provide overlap conditioning to the next segment.

To get the same speed and savings, I refactored the final assembly block to use native JAX on-TPU postprocessing.

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py [hunk 5]
```diff
@@ -1111,64 +1141,98 @@ def _encode(self, x: jax.Array, feat_cache: AutoencoderKLWanCache):
...
+      local_encoder = nnx.merge(graphdef, state)
+      out_chunk, next_feat_map, _ = local_encoder(chunk, feat_cache=current_feat_map, feat_idx=0)
       out_chunk = jax.lax.with_sharding_constraint(out_chunk, spatial_sharding)
+      next_feat_map = jax.tree_util.tree_map(
+          lambda x: jax.lax.with_sharding_constraint(x, spatial_sharding) if isinstance(x, jax.Array) else x, next_feat_map
+      )
       return next_feat_map, out_chunk
 
     enc_feat_map, out_rest = jax.lax.scan(scan_fn, enc_feat_map, x_scannable)
     # out_rest shape: (iter_-2, B, T', H, W, C) -> transpose back
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/wan/autoencoder_kl_wan.py:1220]: I am doing this a few lines later

File: src/maxdiffusion/pipelines/wan/wan_pipeline.py [hunk 6]
```diff
@@ -648,21 +659,27 @@ def prepare_latents_i2v_base(
...
+      video = (video * 255.0).astype(jnp.uint8)
+      video.block_until_ready()
+    if trace is not None:
+      trace["vae_decode_tpu"] = time.perf_counter() - t_vae_tpu_start
 
-    video = jnp.transpose(video, (0, 4, 1, 2, 3))
     video = jax.experimental.multihost_utils.process_allgather(video, tiled=True)
-    video = torch.from_numpy(np.array(video.astype(dtype=jnp.float32))).to(dtype=torch.bfloat16)
-    return self.video_processor.postprocess_video(video, output_type="np")
+    video = np.array(video)
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/pipelines/wan/wan_pipeline.py:685]: Yes, downstream changes have been made

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py [hunk 7]
```diff
@@ -1111,64 +1141,98 @@ def _encode(self, x: jax.Array, feat_cache: AutoencoderKLWanCache):
...
 
-    spatial_sharding = NamedSharding(self.mesh, P(None, None, None, "vae_spatial", None))
+    spatial_sharding = NamedSharding(self.mesh, P("redundant", None, None, "vae_spatial", None))
 
     def finalize(out, enc_feat_map):
       feat_cache._enc_feat_map = enc_feat_map
       enc = self.quant_conv(out)
-      mu, logvar = enc[:, :, :, :, : self.z_dim], enc[:, :, :, :, self.z_dim :]
-      enc = jnp.concatenate([mu, logvar], axis=-1)
       feat_cache.init_cache()
```
Comment (ninatu) [inline src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: is feat_cache.init_cache() needed here? feat_cache.init_cache() is always called in the beginning of the function, therefore it seems like here it should not be needed

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py [hunk 8]
```diff
@@ -1111,64 +1141,98 @@ def _encode(self, x: jax.Array, feat_cache: AutoencoderKLWanCache):
...
     t = x.shape[1]
-    iter_ = 1 + (t - 1) // 4
+    CHUNK_SIZE = self.vae_encode_chunk
+    if CHUNK_SIZE == -1:
+      CHUNK_SIZE = max(1, t - 1)
+      assert (
+          CHUNK_SIZE % self.temporal_downsample_factor == 0
+      ), f"When vae_encode_chunk is -1, the input sequence length - 1 ({CHUNK_SIZE}) must be a multiple of the temporal downsampling factor ({self.temporal_downsample_factor})."
+    # Number of chunk iterations: 1 for init frame, then (t-1)//CHUNK_SIZE for the rest
+    iter_ = 1 + ((t - 1) // CHUNK_SIZE) if t > 1 else 1
```
Comment (ninatu) [inline src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: Sorry, just double checked, it seems like it should be `1 +  int(math.ceil((t - 1) / CHUNK_SIZE))`?  For the edge case, when the last chunk is not full, right?

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py
Comment (ninatu) [inline src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: It seems like feat_cache.init_cache() is unnecessary here as well

**Review by ninatu — DISMISSED** [review]

File: src/maxdiffusion/tests/wan_kv_cache_test.py [hunk 9]
```diff
@@ -210,7 +210,7 @@ def mock_load_scheduler(config):
     self.assertEqual(video_with_cache[0].shape, (num_frames, height, width, 3))
 
     # Compare outputs
-    np.testing.assert_allclose(video_no_cache, video_with_cache, rtol=1e-1, atol=0.7)
+    np.testing.assert_allclose(video_no_cache, video_with_cache, rtol=1e-1, atol=180)
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/tests/wan_kv_cache_test.py]: Refactored KVCache test to check MAE instead of checking each pixel difference.

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py (hunk 8 above)
Comment (Perseus14) [reply src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: Ahh that's right. I will revert to the previous logic instead of importing a new math package

`iter_ = 1 + ((t - 1 + CHUNK_SIZE - 1) // CHUNK_SIZE) if t > 1 else 1`

Comment (ninatu) [reply src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: could you maybe add the comment about it? Since it was unclear that  `CHUNK_SIZE - 1` addition for rounding up,  I was confused.

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py (hunk 7 above)
Comment (Perseus14) [reply src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: @eltsai Could you confirm? I tested by removing it and it works, video is generated with no quality difference.

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py [hunk 10]
```diff
@@ -159,26 +157,30 @@ def __init__(
...
   ):
     broadcastable_dims = (1, 1, 1) if not images else (1, 1)
     shape = (dim, *broadcastable_dims) if channel_first else (dim,)
     self.eps = eps
     self.channel_first = channel_first
     self.scale = dim**0.5
+    self.dtype = dtype
     # Initialize gamma as parameter
-    self.gamma = nnx.Param(jnp.ones(shape))
+    self.gamma = nnx.Param(jnp.ones(shape, dtype=weights_dtype))
```
Comment (eltsai) [inline src/maxdiffusion/models/wan/autoencoder_kl_wan.py:170]: this is great!

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py [hunk 11]
```diff
@@ -1111,64 +1139,97 @@ def _encode(self, x: jax.Array, feat_cache: AutoencoderKLWanCache):
...
       x = jnp.transpose(x, (0, 2, 3, 4, 1))
       assert x.shape[-1] == 3, f"Expected input shape (N, D, H, W, 3), got {x.shape}"
+    x = x.astype(self.encoder.conv_in.conv.dtype)
 
     t = x.shape[1]
-    iter_ = 1 + (t - 1) // 4
+    CHUNK_SIZE = self.vae_encode_chunk
+    if CHUNK_SIZE == -1:
+      CHUNK_SIZE = max(1, t - 1)
+      assert (
```
Comment (eltsai) [inline src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: Instead of throwing out an exception here, maybe we can do some padding here to make sure it is dividable? Will see if others think this is reasonable.

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py [hunk 12]
```diff
@@ -510,21 +519,23 @@ def __init__(
...
-    # qkv = qkv.reshape(batch_size * time, 1, channels * 3, -1)
-    qkv = qkv.reshape(batch_size * time, 1, -1, channels * 3)
-    qkv = jnp.transpose(qkv, (0, 1, 3, 2))
-    q, k, v = jnp.split(qkv, 3, axis=-2)
-    q = jnp.transpose(q, (0, 1, 3, 2))
-    k = jnp.transpose(k, (0, 1, 3, 2))
-    v = jnp.transpose(v, (0, 1, 3, 2))
+    qkv = self.to_qkv(x)  # Output: (N*T, H, W, C * 3)
+    qkv = qkv.reshape(batch_size * time, height * width, 3, channels)
+    q = qkv[:, :, 0, :]  # (B*T, H*W, C)
```
Comment (eltsai) [inline src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: Curious if we have numbers showing the exact speedup from this block alone?

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py (hunk 11 above)
Comment (Perseus14) [reply src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: Done!

File: src/maxdiffusion/models/wan/autoencoder_kl_wan.py (hunk 12 above)
Comment (Perseus14) [reply src/maxdiffusion/models/wan/autoencoder_kl_wan.py]: I tested it and found that the original implementation is much better interms of speed and memory. I have reverted it to the original code

**Review by eltsai — APPROVED** [review]
